src/Features/CSharp/Portable/SignatureHelp/AbstractCSharpSignatureHelpProvider.cs

```diff
@@ -72,5 +80,288 @@ protected IList<TaggedText> GetAwaitableUsage(IMethodSymbol method, SemanticMode
+        {
+            if (arguments.Count != 0)
+            {
+                foreach (var method in methodGroup)
+                {
+                    if (IsAcceptable(arguments, method, position, semanticModel, semanticFactsService, out var parameterIndex))
+                    {
+                        return (method, parameterIndex);
+                    }
+                }
+            }
+
+            // Note: Providing no recommendation if no arguments allows the model to keep the last implicit choice
+            return (null, -1);
+        }
+
+        /// <summary>
+        /// If an argument name does not correspond to a parameter name, this method is unacceptable. We'll filter it out.
+        /// </summary>
+        protected static bool IsUnacceptable(SeparatedSyntaxList<ArgumentSyntax> arguments, IMethodSymbol method)
+        {
+            if (arguments.Count == 0)
+            {
+                return false;
+            }
+
+            var parameterNames = method.Parameters.SelectAsArray(p => p.Name);
+            foreach (var argument in arguments)
+            {
+                if (HasName(argument, out var name))
+                {
+                    if (!parameterNames.Contains(name))
+                    {
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        protected static bool IsAcceptable(SeparatedSyntaxList<ArgumentSyntax> arguments, IMethodSymbol method, int position,
+            SemanticModel semanticModel, ISemanticFactsService semanticFactsService, out int foundParameterIndex)
+        {
+            int argumentCount = arguments.Count;
+            var parameters = method.Parameters;
+            var parameterCount = parameters.Length;
+
+            // map the arguments to their corresponding parameters
+            var map = ArrayBuilder<int>.GetInstance(argumentCount, -1);
+            if (!PrepareMap())
+            {
+                foundParameterIndex = -1;
+                return false;
+            }
+
+            // verify that the arguments are compatible with their corresponding parameters
+            for (var argumentIndex = 0; argumentIndex < argumentCount; argumentIndex++)
+            {
+                var parameterIndex = map[argumentIndex];
+                if (parameterIndex < 0)
+                {
+                    continue;
+                }
+
+                var parameter = parameters[parameterIndex];
+                var argument = arguments[argumentIndex];
+
+                if (!IsAcceptableArgument(argument, parameter))
+                {
+                    foundParameterIndex = -1;
+                    return false;
+                }
+            }
+
+            // find the parameter at the cursor position
+            var argumentIndexToSave = GetArgumentIndex(arguments, position);
```

Comment:
skipped from this point on.  looking for some docs/overall-comments explaining what's going on.